### PR TITLE
Adding variables to event

### PR DESCRIPTION
**Pull request by anmalara**
Redefiniton of minimum event variables
Add variables to pfcands
Add jet index to pfcands
Fixed convention for getters
Moved jets to ak4chs
Add jet collections
Updated jet ID
Added E/gamma energy fractions (EM fraction is deprecated. to be removed eventually)
Added year var to event ( to be tested as it doesn't show up in the tree, like weight)

### CODE REVIEW

File: Analyzer/plugins/NTuplizer.cc [hunk 1]
```diff
@@ -225,48 +246,52 @@ bool NTuplizer::filter(edm::Event& iEvent, const edm::EventSetup& iSetup){
...
+    event.weight = geninfo->weight(); // equivalent to geninfo->weights().at(0), always returns the 0th element
+    // Reco-event variables based on simulation
+    event.ntrueint = pus->at(0).getTrueNumInteractions();
+    event.npu = 0;
+    for(size_t i=0; i<pus->size(); ++i){
+      if(pus->at(i).getBunchCrossing() == 0){ // intime pileup
+        event.npu += pus->at(i).getPU_NumInteractions();
+      }
     }
+  } else{
```
Comment (reimersa) [inline Analyzer/plugins/NTuplizer.cc:285]:
```
}
else{
```

File: Analyzer/plugins/NTuplizer.cc [hunk 2]
```diff
@@ -163,10 +176,11 @@ NTuplizer::NTuplizer(const edm::ParameterSet& iConfig){
 
 bool NTuplizer::filter(edm::Event& iEvent, const edm::EventSetup& iSetup){
 
-
   // declare handles
   edm::Handle<std::vector<pat::Muon>> muons;
-  edm::Handle<std::vector<pat::Jet>> jets;
+  edm::Handle<std::vector<pat::Jet>> ak4chs;
+  edm::Handle<std::vector<pat::Jet>> ak4puppi;
+  edm::Handle<std::vector<pat::Jet>> ak8puppi;
```
Comment (reimersa) [inline Analyzer/plugins/NTuplizer.cc:183]: would be nice to rename also these to the new naming a la `jets_ak4chs`.

File: Analyzer/plugins/NTuplizer.cc [hunk 3]
```diff
@@ -132,8 +143,10 @@ NTuplizer::NTuplizer(const edm::ParameterSet& iConfig){
     tree->Branch("Events", &event);
   }
 
+  token_ak4chs           = consumes<std::vector<pat::Jet>> (iConfig.getParameter<edm::InputTag>("ak4chs"));
+  token_ak4puppi         = consumes<std::vector<pat::Jet>> (iConfig.getParameter<edm::InputTag>("ak4puppi"));
+  token_ak8puppi         = consumes<std::vector<pat::Jet>> (iConfig.getParameter<edm::InputTag>("ak8puppi"));
```
Comment (reimersa) [inline Analyzer/plugins/NTuplizer.cc:148]: would be nice to rename also these to the new naming a la jets_ak4chs.

File: Analyzer/plugins/NTuplizer.cc [hunk 4]
```diff
@@ -949,18 +905,136 @@ bool NTuplizer::filter(edm::Event& iEvent, const edm::EventSetup& iSetup){
...
+      if(patcand.hasTrackDetails()){
+        p.set_dz_err(patcand.dzError());
+        p.set_dxy_err(patcand.dxyError());
+      }
+      p.set_vertex_x(patcand.vertex().X());
+      p.set_vertex_y(patcand.vertex().Y());
+      p.set_vertex_z(patcand.vertex().Z());
+      p.set_is_iso_ch_had(patcand.isIsolatedChargedHadron());
+
+      if(true) {
```
Comment (reimersa) [inline Analyzer/plugins/NTuplizer.cc:921]: This doesn't look particularly useful

File: Analyzer/plugins/NTuplizer.cc [hunk 5]
```diff
@@ -949,18 +905,136 @@ bool NTuplizer::filter(edm::Event& iEvent, const edm::EventSetup& iSetup){
...
+        for(size_t jetidx_=0; jetidx_<ak4chs->size(); jetidx_++) {
+          if(!ak4chs->at(jetidx_).isPFJet()) continue;
+          pat::Jet patjet = ak4chs->at(jetidx_);
+          const auto& jet_daughter_ptrs = patjet.daughterPtrVector();
+          for(const auto & daughter_p : jet_daughter_ptrs){
+            auto r = reco::deltaR(p.eta(), p.phi(), daughter_p->eta(), daughter_p->phi());
+            if (closeFloat(r, 0.0f) && closeFloat(p.pt(), daughter_p->pt())) {
+              jetidx = jetidx_;
+              break;
+            }
```
Comment (reimersa) [inline Analyzer/plugins/NTuplizer.cc:932]: is there a CMSSW function for this?

File: Analyzer/plugins/NTuplizer.cc [hunk 6]
```diff
@@ -949,18 +905,136 @@ bool NTuplizer::filter(edm::Event& iEvent, const edm::EventSetup& iSetup){
...
+    j.set_L1_factor(is_puppi? 1.: patjet.correctedJet("L1FastJet").pt() / patjet.correctedJet("Uncorrected").pt());
+
+    float NHF  = patjet.neutralHadronEnergyFraction();
+    float CHF  = patjet.chargedHadronEnergyFraction();
+    float NEMF = patjet.neutralEmEnergyFraction();
+    float CEMF = patjet.chargedEmEnergyFraction();
+    float MUOF = patjet.muonEnergyFraction();
+    float ELEF = patjet.electronEnergyFraction();
+    float PHOF = patjet.photonEnergyFraction();
+    float sumFrac = NHF+CHF+NEMF+MUOF+CEMF;
```
Comment (reimersa) [inline Analyzer/plugins/NTuplizer.cc:974]: can be removed

File: Analyzer/plugins/NTuplizer.cc [hunk 7]
```diff
@@ -949,18 +905,136 @@ bool NTuplizer::filter(edm::Event& iEvent, const edm::EventSetup& iSetup){
...
+    int CHM = patjet.chargedMultiplicity();
+    int NM = (!is_puppi || is_ak8)? patjet.neutralMultiplicity(): patjet.userFloat("patPuppiJetSpecificProducer:neutralPuppiMultiplicity");
+    int NumConst = CHM+NM;
+    j.set_n_constituents(NumConst);
```
Comment (reimersa) [inline Analyzer/plugins/NTuplizer.cc:978]: reminder: check if this is a sensible number, I seem to remember a bug in older ntuples

File: Analyzer/plugins/NTuplizer.cc [hunk 8]
```diff
@@ -949,18 +905,136 @@ bool NTuplizer::filter(edm::Event& iEvent, const edm::EventSetup& iSetup){
...
+        if (is_puppi){ pass_tight = NHF < 0.9999;}
+        else {pass_tight = (NEMF > 0.0 && NEMF < 0.99 && NM > 1);}
+      }
+      pass_tight_lepveto = pass_tight;
+    }
+    else if(fabs(patjet.eta()) <= 5.0){
+      if (is_puppi) {pass_tight = (NEMF < 0.9 && NM > 2);}
+      else {pass_tight = (NHF > 0.2 && NEMF < 0.9 && NM > 10);}
+      pass_tight_lepveto = pass_tight;
+    }
```
Comment (reimersa) [inline Analyzer/plugins/NTuplizer.cc:1017]: format to either
```
if(){
//
}
else{
//
}
```

or 
```
if() //
else //
```

File: Analyzer/plugins/NTuplizer.cc (hunk 5 above)
Comment (anmalara) [reply Analyzer/plugins/NTuplizer.cc:932]: no idea. cannot find anything

File: Analyzer/plugins/NTuplizer.cc (hunk 4 above)
Comment (anmalara) [reply Analyzer/plugins/NTuplizer.cc:921]: removed. just for reference, 
```
matchByCommonSourceCandidatePtr(patcand, jet)
```
returns always false. Not sure why

**Comment by reimersa** [comment]
thanks for the small changes, good to go!

**Pull request merged**